src/libaducpal/src/ftw.c

```diff
@@ -165,318 +78,124 @@ static int construct_typeflag(const DWORD attrib, int flags)
-            {
-                /* Check if no files match pattern. */
-                if (GetLastError() == ERROR_FILE_NOT_FOUND)
-                    goto subdirs;
-
-                /* Bail out from other errors. */
-                dir->path[dir->pathlen] = '\0';
-                // oops(dir->path);
-                rval = 1;
-                goto popdir;
-            }
+        if (entry->d_type == DT_DIR)
+        {
+            ret = 0;
+
+            const size_t buffer_len = ftwbuf.base + 1 + entry->d_namlen + 1;
+            char* subdir = malloc(buffer_len);
+            sprintf_s(subdir, buffer_len, "%s/%s", path, entry->d_name);
 
-            /* Remove pattern from path; keep backslash. */
-            dir->path[dir->slashlen] = '\0';
+            // FTW_DEPTH
+            // If set, do a post-order traversal, that is, call fn() for the directory itself after handling the contents of the
+            //  directory and its subdirectories. (By default, each directory is handled before its contents.)
 
-            /* Print all files to match pattern. */
-            do
+            if (flags & FTW_DEPTH)
             {
-                const DWORD attr = dir->ffd.dwFileAttributes;
-                int typeflag = construct_typeflag(attr, flags);
-
-                struct FTW lev = { 0 };
-
-                struct _stat st64i32;
-                _wstat(wpath, &st64i32);
-
-                // TODO: Convert to stat. (Only st_dev and st_ino are used, however)
-                struct stat st;
-                memset(&st, sizeof(st), 0);
-                st.st_dev = st64i32.st_dev;
-                st.st_ino = st64i32.st_ino;
-                st.st_mode = st64i32.st_mode;
-                st.st_mode = st64i32.st_mode;
-                st.st_nlink = st64i32.st_nlink;
-                st.st_uid = st64i32.st_uid;
-                st.st_gid = st64i32.st_gid;
-                st.st_rdev = st64i32.st_rdev;
-                st.st_size = st64i32.st_size;
-                st.st_atime = st64i32.st_atime;
-                st.st_mtime = st64i32.st_mtime;
-                st.st_ctime = st64i32.st_ctime;
-
-                struct history new;
-                new.chain = h;
-                new.dev = st.st_dev;
-                new.ino = st.st_ino;
-                new.level = h ? h->level + 1 : 0;
-
-                /* TODO not recursive, so refactor this to make lev be
-				 * function-scope and increment/decerment lev.level as
-				 * directories are entered/exited; it will also know to
-				 * exit when fd_limit is exceeded
-				 */
-                (void)fd_limit;
-                lev.level = new.level;
-                lev.base = (int)wcslen(dir->path);
-
-                const wchar_t* thisname = dir->ffd.cFileName;
-
-                /* skip directories named '.' and '..' */
-                if (thisname[0] == L'.' && (!thisname[1] || (thisname[1] == L'.' && !thisname[2])))
+                ret = do_nftw(subdir, func, nopenfd, flags, level + 1);
```

Comment:
recursive is not ideal. a non-recursive approach that queues subdirs would be better approach to keep the callstack depth shallower.

Reply:
Perhaps -- do you know how large (in K) a typical Linux callstack is?  I've used this about 8 levels deep and didn't run into any issues.